teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,330 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+- This is adding complexity to both the implementations and the testing matrix for the newly introduced flag.
+
+
+### Make validation changes only for new beta features
+This alternative proposes keeping beta features on by default. It will keep the current implementations and validations for beta features in `v1` and `v1beta1` apiVersion and only make new features have the synchronized validations across different apiVersions. This also proposes to keep the default value of `enable-api-fields` as `beta`,
+
+#### Pros
+- This is backwards compatible.
+- Changes are minimal for the current codebase with probably only documentation required.
+#### Cons
+- We would still need to use `beta` as the default stability of features that are turned on. Some beta features are still coupled in feature and API versioning. This will not allow cluster operators to opt-in only stable features in v1beta1.
+
+### Give 9-month warning before breaking changes and default to stable
+This alternative proposes the validation change for stable features in v1beta1 to be done with giving a notice for the breaking change for 9 months.
+
+#### Cons:
+- Since v1beta1 only has 12 months of support period left, making a breaking change to it might have more impacts on v1beta1 users than its worth after 9 months.
+
+## Implementation Plan
+
+For each new feature driven by api fields, a new feature flag using FeatureFlag struct will be added to the existent set of feature flags. This struct will include the stability of feature and whether it is turned on or locked to default. Currently it will only work for new features without `enable-api-fields`, see future work for more details of the option to introduce `enable-api-fields=none` to use per feature flags for new features.
+
+
+```go
+type FeatureFlag struct {
+    // Name of the feature flag
+    Name string
+    
+    // Stability level of the feature, one of "stable", "beta" or "alpha"
+    Stability string
+    
+    // Enabled is whether the feature is turned on
+    Enabled bool
+    
+    // LockToDefault indicates whether the feature is locked to default value and cannot be changed
+    LockToDefault bool
+    
+    // Deprecated is whether the feature is deprecated
+    // +optional
+    Deprecated bool
+}
+```
+
+## Test Plan
+### Retain existing CI end-to-end testing matrix 
+We plan to retain the existing testing matrix for fields at `alpha`, `beta` and `stable`. With the addition of per-feature flags for new features, it would look like:
+
+| |enable-api-fields|Per feature flags|Integration tests|
+| -- | -- | -- | -- |
+|Opt-in stable| stable|Turn OFF all per feature flags|Run all stable e2e tests.|
+|Opt-in beta| beta|Turn ON all beta per-feature flags|Run all beta e2e tests.|
+|Opt-in alpha| alpha|Turn ON all per-feature flags (alpha and beta)|Run all alpha e2e tests.|
+
+
+## Additional CI tests
+### Additional Test Combinations
```

Comment:
is this in addition to testing with all alpha/beta/stable features on?

Reply:
Yes, these are for the purpose of testing the combinations of per feature flags.

Reply:
I am having hard time understanding the following table. How do we choose a small number of `e2e` tests? What are we achieving with this kind of testing?

We might have to think this a little more, its not a blocker but something we need to address as part of this proposal.

How is Kubernetes (an inspiration for this proposal) handling it?

Some key points for us to think about:

The cluster operator can enable one or more features using per feature flags. Pipeline/task authors can choose to leverage none or only one or multiple of the enabled features in their specifications. As a project maintainer, we have to make sure a feature works in isolation and also does not impact any existing functionalities.